boto3/docs/utils.py

```diff
@@ -11,11 +11,23 @@
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
 import inspect
+import re
 from botocore.compat import six
 
+TARGET_COMPONENT_RE = re.compile(r'[^.\[\]]+(?![^\[]*\])')
+
 
 def get_resource_ignore_params(params):
```

Comment:
FWIW, the sub expressions are left associative so you just need to grab the left most leaf node in the AST which will be the first field name.  Something like this (probably want some error checking to make sure that node is a `field` type):

```
>>> result = jmespath.compile('foo[].bar[].baz.*.qux')
>>> current = result.parsed
>>> while current['children']:
...   current = current['children'][0]
...
>>> current['value']
'foo'
```